src/thermo/Phase.cpp

```diff
@@ -256,42 +256,122 @@ std::string Phase::speciesSPName(int k) const
+        } else {
+            return {"T", "P"};
+        }
+    } else {
+        if (isCompressible()) {
+            return {"T", "density", "Y"};
+        } else {
+            return {"T", "P", "Y"};
+        }
+    }
+}
+
+vector<std::string> Phase::fullStates() const
+{
+    if (isPure()) {
+        if (isCompressible()) {
+            return {"TD", "TP", "UV", "DP", "HP", "SP", "SV"};
+        } else {
+            return {"TP", "HP", "SP"};
+        }
+    } else {
+        if (isCompressible()) {
+            return {"TDX", "TDY", "TPX", "TPY", "UVX", "UVY", "DPX", "DPY",
+                    "HPX", "HPY", "SPX", "SPY", "SVX", "SVY"};
+        } else {
+            return {"TPX", "TPY", "HPX", "HPY", "SPX", "SPY"};
+        }
+    }
+}
+
+vector<std::string> Phase::partialStates() const
+{
+    if (isPure()) {
+        return {};
+    } else {
+        if (!isCompressible()) {
+            return {"TP", "HP", "SP"};
+        } else {
+            return {"TD", "TP", "UV", "DP", "HP", "SP", "SV"};
+        }
+    }
+}
+
 void Phase::saveState(vector_fp& state) const
 {
-    state.resize(nSpecies() + 2);
+    if (isPure()) {
+        state.resize(2);
+    } else {
+        state.resize(nSpecies() +2);
+    }
     saveState(state.size(), &state[0]);
 }
+
 void Phase::saveState(size_t lenstate, doublereal* state) const
 {
     state[0] = temperature();
-    state[1] = density();
-    getMassFractions(state + 2);
+    if (isCompressible()) {
+        state[1] = density();
+    } else {
+        state[1] = pressure();
+    }
+    if (!isPure()) {
+        if (defaultState()[2] == "Y") {
+            getMassFractions(state + 2);
+        } else {
+            getMoleFractions(state + 2);
+        }
+    }
 }
 
 void Phase::restoreState(const vector_fp& state)
 {
     restoreState(state.size(),&state[0]);
-    compositionChanged();
 }
 
-void Phase::restoreState(size_t lenstate, const doublereal* state)
+void Phase::restoreState(size_t lenstate, const double* state)
 {
-    if (lenstate >= nSpecies() + 2) {
-        setMassFractions_NoNorm(state + 2);
-        setTemperature(state[0]);
+    size_t ls = 2;
+    if (!isPure()) {
+        ls += nSpecies();
+    }
+    if (lenstate < ls) {
+        throw ArraySizeError("Phase::restoreState",
+                             lenstate, ls);
+    }
+
+    setTemperature(state[0]);
+    if (isCompressible()) {
         setDensity(state[1]);
     } else {
-        throw ArraySizeError("Phase::restoreState",
-                             lenstate,nSpecies()+2);
+        setPressure(state[1]);
+    }
+
+    if (!isPure()) {
+        if (defaultState()[2] == "Y") {
+            setMassFractions_NoNorm(state + 2);
+        } else {
+            setMoleFractions_NoNorm(state + 2);
+        }
     }
+    compositionChanged();
 }
 
-void Phase::setMoleFractions(const doublereal* const x)
+void Phase::setMoleFractions(const double* const x)
 {
+    assertNotPure("setMoleFractions");
```

Comment:
I think the change to `PureFluid` and others to not need a call to `setMoleFractions` to update `m_mmw` is fine. But I don't see a reason why, at runtime, code can't use the `setMoleFractions` for a pure substance. It is effectively a no-op and allowing this to happen simplifies the mental pattern that people have to carry around about how states can be set in Cantera.

Reply:
All checks are removed.